cuda/gemv_kernels.cuh

```diff
@@ -0,0 +1,184 @@
+#pragma once
+
+#include <cooperative_groups.h>
+#include <cublas_v2.h>
+
+
+#include <accessor/range.hpp>
+#include <accessor/reduced_row_major.hpp>
+#include <cinttypes>
+
+
+#include "kernel_utils.cuh"
+#include "utils.cuh"
+
+
+namespace kernel {
+
+
+namespace cg = cooperative_groups;
+
+template <std::int64_t block_size, typename ValueType>
+__global__ __launch_bounds__(block_size) void gemv(
+    const matrix_info minfo, ValueType alpha, const ValueType *__restrict__ mtx,
+    const matrix_info x_info, const ValueType *__restrict__ x,
+    const matrix_info res_info, ValueType beta, ValueType *__restrict__ res)
+{
+    // expect x_info.size[1] == 1
+    const std::int64_t row_idx{blockIdx.x};
+    if (row_idx >= minfo.size[0]) {
+        return;
+    }
+    const std::int64_t row_start = row_idx * minfo.stride;
+    // can't use array with `error_type`
+    __shared__ char shared_impl[sizeof(ValueType) * block_size];
+    auto shared = reinterpret_cast<ValueType *>(shared_impl);
+
+    ValueType local_result{};
+    const auto group = cg::this_thread_block();
+    const auto local_id = group.thread_rank();
+
+    for (std::int64_t col = local_id; col < minfo.size[1]; col += block_size) {
+        const auto mtx_val = mtx[row_start + col];
+        const auto b_val = x[col * x_info.stride];
+        local_result += mtx_val * b_val;
+    }
+    shared[local_id] = local_result;
+    reduce(group, shared, [](ValueType a, ValueType b) { return a + b; });
+    if (local_id == 0) {
+        const auto res_idx = row_idx * res_info.stride;
+        if (beta == ValueType{0}) {
+            res[res_idx] = alpha * shared[local_id];
+        } else {
+            res[res_idx] = alpha * shared[local_id] + beta * res[res_idx];
+        }
+    }
+}
+
+template <std::int64_t block_size, typename MtxRange, typename XRange,
+          typename ResRange, typename ArType>
+__global__ __launch_bounds__(block_size) void acc_gemv(ArType alpha,
+                                                       MtxRange mtx, XRange x,
+                                                       ArType beta,
+                                                       ResRange res)
+{
+    using ar_type = decltype(mtx(0, 0) + mtx(0, 0));
```

Comment:
same question as dot. shouldn't it be `alpha * mtx(0, 0) * x(0, 0) + beta * res(0, 0)`?